### PR TITLE
Add simulator for NXT

### PR DESCRIPTION
**Pull request by mandaiy**
This commit adds a feature of simulating NXT robots.
Users of the simulator can use NxtMachineSimulator by instantiating that class.
The simulator class allows users to create _simulatable_ version of sensors/motors by the same way
as normal NxtMachine.

### CODE REVIEW

**Comment by mandaiy** [comment]
According to #12, I added a simulator feature for NXT robots.
One thing I misunderstood is that users manipulate robots via Sensors/Motors.
So I implemented some _simulatable_ sensors/motors.
Stay tuned for Ev3 and Pile version.

TODO: add test cases, general controllers.

Please feel free to leave any opinions.

**Comment by mandaiy** [comment]
Added a simple use-case in [drive](https://github.com/myusak/drive/tree/feature/simulator).
You can run a program on your emulator.

**Comment by tiwanari** [comment]
> One thing I misunderstood is that users manipulate robots via Sensors/Motors.
> So I implemented some simulatable sensors/motors.

I see. The workaround looks good to keep loose couplings :)

File: src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package com.pileproject.drivecommand.machine.device.input.simulation;
```
Comment (makotoshimazu) [inline src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java:1]: I feel these objects are just for mocking. What do you think about moving to com.pileproject.drivecommand.**mock**.machine.device.input to clarify which classes are available for testing?

File: src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+import com.pileproject.drivecommand.machine.device.input.ColorSensor;
+import com.pileproject.drivecommand.machine.device.port.InputPort;
+import com.pileproject.drivecommand.model.ProtocolBase;
+
+public class SimulatedColorSensor extends ColorSensor {
```
Comment (makotoshimazu) [inline src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java:7]: I feel Mock is more common word for this purpose. How about MockColorSensor?

File: src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    private float red;
+    private float green;
+    private float blue;
+
+    private int illuminance;
+
+    public SimulatedColorSensor(InputPort port, ProtocolBase protocol) {
+        super(port, protocol);
```
Comment (makotoshimazu) [inline src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java:16]: Is super necessary?

File: src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import com.pileproject.drivecommand.machine.device.port.InputPort;
+import com.pileproject.drivecommand.model.ProtocolBase;
+
+public class SimulatedColorSensor extends ColorSensor {
+
+    private float red;
+    private float green;
+    private float blue;
+
+    private int illuminance;
```
Comment (makotoshimazu) [inline src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java:13]: Adding `m` before each members for keeping the naming convention?

File: src/main/java/com/pileproject/drivecommand/model/com/NullCommunicator.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package com.pileproject.drivecommand.model.com;
```
Comment (makotoshimazu) [inline src/main/java/com/pileproject/drivecommand/model/com/NullCommunicator.java:1]: How about moving this class to com.pileproject.drivecommand.mock.model?

File: src/main/java/com/pileproject/drivecommand/model/com/NullCommunicator.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+import java.io.IOException;
+
+/**
+ * Null object of {@link ICommunicator}.
+ */
+public class NullCommunicator implements ICommunicator {
```
Comment (makotoshimazu) [inline src/main/java/com/pileproject/drivecommand/model/com/NullCommunicator.java:8]: Let's rename this to MockCommunicator if you agree with the "mock" prefix.

File: src/main/java/com/pileproject/drivecommand/model/nxt/NxtMachineSimulator.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+import com.pileproject.drivecommand.machine.device.input.TouchSensor;
+import com.pileproject.drivecommand.machine.device.input.simulation.SimulatedLineSensor;
+import com.pileproject.drivecommand.machine.device.input.simulation.SimulatedSoundSensor;
+import com.pileproject.drivecommand.machine.device.input.simulation.SimulatedTouchSensor;
+import com.pileproject.drivecommand.machine.device.output.Motor;
+import com.pileproject.drivecommand.machine.device.output.simulation.SimulatedMotor;
+import com.pileproject.drivecommand.machine.device.port.InputPort;
+import com.pileproject.drivecommand.machine.device.port.OutputPort;
+
+public class NxtMachineSimulator extends NxtMachine {
```
Comment (makotoshimazu) [inline src/main/java/com/pileproject/drivecommand/model/nxt/NxtMachineSimulator.java:15]: I think this class could targets not only NxtMachine but all classes derived from MachineBase.
We can move this class to com.pileproject.drivecommand.mock.model, for example, and can also implement MockMachine inheriting MachineBase.

File: src/main/java/com/pileproject/drivecommand/model/nxt/NxtNullProtocol.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
+package com.pileproject.drivecommand.model.nxt;
+
+import com.pileproject.drivecommand.command.CommandBase;
+import com.pileproject.drivecommand.model.com.NullCommunicator;
+
+import java.io.IOException;
+import java.util.Map;
+
+public class NxtNullProtocol extends NxtProtocol {
```
Comment (makotoshimazu) [inline src/main/java/com/pileproject/drivecommand/model/nxt/NxtNullProtocol.java:9]: This also can be moved to com.pileproject.drivecommand.mock.model.

**Review by makotoshimazu — COMMENTED** [review]
These comments are based on my comment below.
https://github.com/PileProject/drivecommand/issues/12#issuecomment-257579000

File: src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java (hunk 1 above)
Comment (mandaiy) [reply src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java:1]: Hmm, I don't know which is better to be honest.
@tiwanari might have opinions on this.

File: src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java (hunk 2 above)
Comment (mandaiy) [reply src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java:7]: Ok.

File: src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java (hunk 4 above)
Comment (mandaiy) [reply src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java:13]: Yeah yeah gotcha gotcha...

File: src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java (hunk 3 above)
Comment (mandaiy) [reply src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java:16]: It's necessary because the superclass does not have a default constructor and then the subclasses must call one of its superclass's constructors.
Welcome to Java!

File: src/main/java/com/pileproject/drivecommand/model/com/NullCommunicator.java (hunk 5 above)
Comment (mandaiy) [reply src/main/java/com/pileproject/drivecommand/model/com/NullCommunicator.java:1]: Same as the first reply

File: src/main/java/com/pileproject/drivecommand/model/com/NullCommunicator.java (hunk 6 above)
Comment (mandaiy) [reply src/main/java/com/pileproject/drivecommand/model/com/NullCommunicator.java:8]: K.

File: src/main/java/com/pileproject/drivecommand/model/nxt/NxtMachineSimulator.java (hunk 7 above)
Comment (mandaiy) [reply src/main/java/com/pileproject/drivecommand/model/nxt/NxtMachineSimulator.java:15]: Once I thought exactly same thing as you thought.
But users of this library might use only one of the machines in this library and 
in that case the corresponding mocking class has to be on the line of inheritance of the machine.
So we'd better to provide mocks for the every machines.

File: src/main/java/com/pileproject/drivecommand/model/nxt/NxtNullProtocol.java (hunk 8 above)
Comment (mandaiy) [reply src/main/java/com/pileproject/drivecommand/model/nxt/NxtNullProtocol.java:9]: Same as the first reply

File: src/main/java/com/pileproject/drivecommand/model/nxt/NxtMachineSimulator.java (hunk 7 above)
Comment (makotoshimazu) [reply src/main/java/com/pileproject/drivecommand/model/nxt/NxtMachineSimulator.java:15]: In my understanding, MachineBase should be the interface exactly for such purpose when we designed this library. User's program should handle MachineBase in their codes except for the factory class (currently it's NxtController, right?).

> But users of this library might use only one of the machines in this library

I know there is such case, but this library should assume what I mentioned above. Additionally, I'm wondering if there is benefit when users directly use the derived classes like NxtMachine.

Comment (mandaiy) [reply src/main/java/com/pileproject/drivecommand/model/nxt/NxtMachineSimulator.java:15]: Your opinion seems right.
One thing which made me confused was that NxtController takes NxtMachine as its constructor argument, which seems to have to be MachineBase.
And now I feel that the controller's name is misleading as it implies general enough to be able to control any nxt machines.

Comment (mandaiy) [reply src/main/java/com/pileproject/drivecommand/model/nxt/NxtMachineSimulator.java:15]: and NxtController is not a factory class. now MachineProvider is responsible for that.

Comment (makotoshimazu) [reply src/main/java/com/pileproject/drivecommand/model/nxt/NxtMachineSimulator.java:15]: I'm not familiar with the actual codes, but I think NxtController should take NxtMachine because NxtController binds each port, which is Nxt(Input|Output)Port, with each device like motors and sensors. 
Also, why is MachineProvider separated with the corresponding controller? In my understanding, the controller class shapes the machine. In other words, User's programs can define the model of his/her machine by implementing the controller class (ex: the machine should have two motors, and two sensors). This means MachineBase itself without controller is useless for the user's code ([drive](https://github.com/PileProject/drive) in our case). That's why I thought "(currently it's NxtController, right?).".

Comment (mandaiy) [reply src/main/java/com/pileproject/drivecommand/model/nxt/NxtMachineSimulator.java:15]: then the mock class has to be the sub class of the corresponding concrete class doesn't it?
if not, NxtController cannot take mocking instance.

Comment (mandaiy) [reply src/main/java/com/pileproject/drivecommand/model/nxt/NxtMachineSimulator.java:15]: and technically machines and controllers are tied up in the provider class. it's just because for the simplicity of the implementation.

File: src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java (hunk 1 above)
Comment (tiwanari) [reply src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java:1]: So, at first, let me confirm options. They are; 
- `com.pileproject.drivecommand.machine.device.input.mock` (as followed the naming discussion)
- `com.pileproject.drivecommand.mock.machine.device.input`

right?

File: src/main/java/com/pileproject/drivecommand/model/nxt/NxtMachineSimulator.java (hunk 7 above)
Comment (tiwanari) [inline src/main/java/com/pileproject/drivecommand/model/nxt/NxtMachineSimulator.java]: Okay... let me summarize your comments;

### How about making a common Mock class

Yeah, it's a reasonable way. But, what use cases do you imagine? I think we should refer some use cases (or a purpose) for each case (a common Mock class / Mock classes derived from `*Machine`s).

I feel "Mock classes derived from `*Machine`s" is a case that we, [drive](https://github.com/PileProject/drive) developers, come up with at first because we have separated apps and these classes will be used in the corresponding apps which require a concrete machine class as an argument for their controllers (these controllers have some naming problems as @myusak mentioned. please see the next section).

### The name, `NxtController`

Some problems of the name; 
- It doesn't mention about the form of NXT but, in [drive](https://github.com/PileProject/drive), the class specifies the form/sensors/motors as @amiq11 mentioned (these controllers specify the forms). So the name is too general for the purpose from a viewpoint Maybe, we should rename it like `NxtCarController` if we introduce it in this library.
- It indeed says `this class controls an NXT`. So, the class should take a `NxtMachine` or one of its children. This is a case in which we need such mocking classes, not a general mocking class.


### In short...
I think this is a very complicated problem and you both are right. It depends on situations. 


### What we should discuss

- What cases do you suppose with the 2 options? (i.e., what is the purpose of each option?)
- Should we choose one of them? Or can we choose both?
- How much should we do in this library?
  - Assuming all cases is impossible
- Are some design changes necessary ? (In particular, as for arguments etc.)

File: src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java (hunk 1 above)
Comment (tiwanari) [reply src/main/java/com/pileproject/drivecommand/machine/device/input/simulation/SimulatedColorSensor.java:1]: Anyway, I agree with @amiq11's idea in the case we use these classes just for testing because users can easily understand where these mocks are placed. 
Let's regard the `mock` package as a small repository for mocks (It may help some developers who implement a communication for other robots).

If some classes, especially `*Controller`s, in this PR are general enough (i.e., it can be used not only for such testing/mocking but also for other purposes), we shouldn't move them into the `mock` package.

File: src/main/java/com/pileproject/drivecommand/model/nxt/NxtMachineSimulator.java (hunk 7 above)
Comment (tiwanari) [reply src/main/java/com/pileproject/drivecommand/model/nxt/NxtMachineSimulator.java:15]: Okay... let me summarize your comments.

### How about making a common Mock class

Yeah, it's a reasonable way. But, what use cases do you imagine? I think we should refer some use cases (or a purpose) for each case (a common Mock class / Mock classes derived from `*Machine`s).

I feel "Mock classes derived from `*Machine`s" is a case that we, [drive](https://github.com/PileProject/drive) developers, come up with at first because we have separated apps and these classes will be used in the corresponding apps which require a concrete machine class as an argument for their controllers (these controllers have some naming problems as @myusak mentioned. please see the next section).

### The name, `NxtController`

Some problems of the name; 
- It doesn't mention about the form of NXT but, in [drive](https://github.com/PileProject/drive), the class specifies the form/sensors/motors as @amiq11 mentioned (these controllers specify the forms). So the name is too general for the purpose from a viewpoint Maybe, we should rename it like `NxtCarController` if we introduce it in this library.
- It indeed says `this class controls an NXT`. So, the class should take a `NxtMachine` or one of its children. This is a case in which we need such mocking classes, not a general mocking class.

### In short...

I think this is a very complicated problem and you both are right. It depends on situations. 

### What we should discuss
- What cases do you suppose with the 2 options? (i.e., what is the purpose of each option?)
- Should we choose one of them? Or can we choose both?
- How much should we do in this library?
  - Assuming all cases is impossible
- Are some design changes necessary ? (In particular, as for arguments etc.)

**Review by tiwanari — COMMENTED** [review]
I left some comments. We need more discussions.

**Comment by tiwanari** [comment]
Oh, because of #19, this PR has a lot of diffs now...

@mandaiy Please take a look on this.
